### DISCUSSION TITLE
有考虑将 `struct` 模块独立出来，做成一个单独的宏包吗?

### DISCUSSION THREAD

**Discussion by zongpingding** [Sophanatprime/cus#60] [Ideas]
就像 CUS 的用户手册中说的那样：
> CuSTEX 重新实现了标题和目录的命令。其中标题的设置方式是 CTEX 风格的，输出目录则提供了多种设置风格，例如 etoc 宏包和 KOMA-Script 文档类的风格。可以在正文中任意位置使用任意次这些命令.

以及文档中的这里:
> struct 模块提供了创建目录和章节标题的方法。参考了 titlesec，titletoc，ctex-heading，etoc 等宏包的实现，并自动阻止加载这些宏包。

也就是 CUS 打算提供了目录和章节定制的相关接口，而且目前来看已经完成了很多的工作了，其所提供的接口已经能够用于绝大部分的任务了. 

目前能够提供比较完善的:
* 标题格式自定义
* 目录格式自定义

的宏包太少了，基本就只有 `titlesec, titletoc, tocloft` 几个宏包，其它的宏包不是太老就是提供的功能太少. 但是前三者也存在一定的问题: 比如 `titlesec` 宏包，对应的实现中[说的那样](https://github.com/jbezos/titlesec/blob/e5b48d9f273c5d4db720804a5f706998a2b642db/titlesec.sty#L25-L31):

> % Admittedly, the current implementation seems too
% complicated, but that's necessary in order to provide
% certain compatibility with the sections as defined by the
% used class. Other packages opt for providing the sections
% as defined by standard classes ignoring the class; for
% instance sectsty which does a simple task in a simple and
% nice way. However, that was not my goal.

我也感觉这个包的实现过于复杂了，提供的接口也不是那么的 friendly. 比如 [CTAN上的评论](https://ctan.org/pkg/titlesec):

> Figuring out the parameters takes some time, but the flexibility is worth it. When writing a thesis with strict standards for everything such packages are priceless.

> Title format changes work as expected. The \titleformat command has over 6 parameters, which is a bit error-prone, but the terse format setting does the job.

也不知道原作者有没有再次重写的打算, 但是似乎也没必要重写，毕竟他提供的功能已经很多了. 尽管有时你可以不依赖任何的宏包或者是文档类，使用原始命令 `\@sect, \@startsection` 手搓以应付比较简单的情况，但是我觉得这个意义其实不大.

所以我很希望 CUS 能够把 `struct` 模块做成一个独立的宏包，但是还是看整个项目的规划吧，我只是在这里提供一个建议.

**Comment by Sophanatprime** [comment]
可以独立成一个 `cusstruct` 宏包，不过目前它依赖很多 `cus` 和 `lt3ekeys` 的东西，都需要拆分出来，而且这个模块的代码还需要重新整理。现在比较忙，还要等一段时间。

**Reply by zongpingding** [reply]
那就等你有空在弄吧

**Reply by Sophanatprime** [reply]
看了一遍代码发现，总的来说 CUS 的其它模块（不包括库）都是为 `struct` 模块服务的，除了 `index` 模块和 `layout` 模块，所以独立的 `cusstruct.sty` 基本上就相当于没有库的 CUS。现在把 `index` 模块划到库了，而且 `layout` 模块也不影响加载其它宏包（它只加载 `lltjp-geometry, geometry, typearea, fancyhdr` 这四个宏包）。所以现在 CUS 应该是可以“安全地”在其它地方使用了，即应当不影响与标题和目录无关的宏包的使用。这是假定 `cus.sty` 和 `cusstruct.sty` 可以同时加载。

如果不需要 `cus.sty` 和 `cusstruct.sty` 可以同时加载，那么只把标题和目录需要的一部分代码打包成 `cusstruct.sty` 即可。如果实在有需要再做吧。
